### PR TITLE
Add IAdapterRegistry.subscribed and Components.rebuildUtilityRegistryFromLocalCache

### LINKED ISSUE #230: Add AdapterRegistry.subscribed()
This would be analogous to `AdapterRegistry.registered()`.

My particular use case for this for consistency checking and error correction. When you call `Components.registerUtility()` , the utility is supposed to wind up in three places:

- Components._utility_registrations
- Components.adapters._adapters
- Components.adapters._subscribers

But we had a situation where a bug (in a subclass) caused only the first and third of those (probably!) to contain the utility. Correcting `_adapters` was easy, because we could check with `registered()` whether it was present or not. Correcting `_subscribers` required implementing a `subscribed()` method.

If we can get a `subscribed()` method, I could also add the error-detecting/correcting method we wrote to `Components`, if desired.

### PR DESCRIPTION
**Pull request by jamadden**
I'm very open to improved naming suggestions for `rebuildUtilityRegistryFromLocalCache`.

Fixes #230

### CODE REVIEW

File: src/zope/interface/adapter.py [hunk 1]
```diff
@@ -430,6 +430,26 @@ def subscribe(self, required, provided, value):
 
         self.changed(self)
 
+    def subscribed(self, required, provided, subscriber):
```
Comment (jamadden) [inline src/zope/interface/adapter.py:433]: This method is almost identical to `registered`, except for which list it operates on (`_adapters` or `_subscribers`), the hardcoded name, and the last line. 

Now that I see that, it's probably worth refactoring to a single implementation.

**Comment by jamadden** [comment]
Hi @mgedmin ! You 👍 the original issue (#230). Is this what you were looking for?

File: src/zope/interface/interfaces.py [hunk 2]
```diff
@@ -997,19 +997,62 @@ def subscribe(required, provided, subscriber): # pylint:disable=arguments-differ
            Subscribers have no names.
         """
 
+    def subscribed(required, provided, subscriber):
+        """
+        Check whether the object *subscriber* is registered directly
+        with this object via a previous
+        call to ``subscribe(required, provided, subscriber)``.
```
Comment (mgedmin) [inline src/zope/interface/interfaces.py:1004]: Word-wrapping feels a little bit weird here.  Any particular reason?

File: src/zope/interface/registry.py [hunk 3]
```diff
@@ -505,6 +505,72 @@ def unregisterHandler(self, factory=None, required=None, name=u''):
...
+        By default, this method only checks for corruption. To make it
+        actually rebuild the registry, pass `True` for *rebuild*.
+
+        :param bool rebuild: If set to `True` (not the default),
+           this method will actually register and subscribe utilities
+           in the registry as needed to synchronize with the local cache.
+
+        :return: A dictionary that's meant as diagnostic data. The keys
+           and values may change over time. When called with a false *rebuild*,
+           the keys ``"needed_registered"`` and ``"needed_subscribed`` will be
```
Comment (mgedmin) [inline src/zope/interface/registry.py:526]:
```suggestion
           the keys ``"needed_registered"`` and ``"needed_subscribed"`` will be
```

**Review by mgedmin — COMMENTED** [review]
I cannot say that I understand what is happening here, but I see nothing obviously wrong with it.

File: src/zope/interface/interfaces.py (hunk 2 above)
Comment (jamadden) [reply src/zope/interface/interfaces.py:1004]: The double-backtick block wraps as a single entity.

File: src/zope/interface/interfaces.py [hunk 4]
```diff
@@ -997,19 +997,62 @@ def subscribe(required, provided, subscriber): # pylint:disable=arguments-differ
...
+
+        If the *subscriber*, or one equal to it, has been subscribed,
+        for the given *required* sequence and *provided* interface,
+        return that object. (This does not guarantee whether the *subscriber*
+        itself is returned, or an object equal to it.)
+
+        If it has not, return ``None``.
```
[the remark this answers is not in the document]
Comment (jamadden) [reply src/zope/interface/interfaces.py:1011]: The new `subscribed()` call is consistent with the existing `registered()` call.

File: src/zope/interface/registry.py (hunk 3 above)
Comment (jamadden) [reply src/zope/interface/registry.py:526]: Thanks!

File: src/zope/interface/interfaces.py (hunk 2 above)
Comment (mgedmin) [reply src/zope/interface/interfaces.py:1004]: Yes, but `call to` is outside the double-ticked block and would easily fit on the previous line.

Comment (jamadden) [reply src/zope/interface/interfaces.py]: I just realized that's what you meant and was committing a change as you commented :)

**Comment by jamadden** [comment]
> I see nothing obviously wrong with it.

I think I'll take that as approval and merge this in the next hour or so unless there's more feedback 😄 

Thank you for the review.

**Pull request merged**